### DISCUSSION TITLE
Solar heating setup questions

### DISCUSSION THREAD

**Discussion by CodeSlinger69** [tagyoureit/nodejs-poolController#808] [Nixie] [answered]
I'm back with you another request for guidance. I had a bit of a delay as I was tracking down a [logic bug](https://github.com/tagyoureit/nodejs-poolController/pull/807), but now I'm on to the configuration of solar heating in Nixie, and as usual, I have questions. I guess the good thing is that I am documenting all these Q&As in the [other thread](https://github.com/tagyoureit/nodejs-poolController/discussions/801), so hopefully this will help others later.

Good news is that I have my solar valve actuator wired into the Sequent Relay Hat, REM can see it, and I can change the valve's position by tapping the valve button in REM.

What I'm hoping to do is two things:
1. Schedule the solar to run periodically, perhaps to heat the pool in early/late season, or to nocturnally cool the pool in the middle season.
2. Allow for the Solar button on dashPanel to be tapped to manually get water flowing to the roof.

At a later point I would like to automatically detect when to heat/cool the pool, but to my understanding requires an additional temperature sensor on the roof, which I don't have. So that's a future project.

I also need the pump to run at a higher RPM to be able to push the water to the roof.

**My question is this:** 
Do I setup "Solar" as a 1) Generic Feature in the Circuits tab, 2) a Solar Heater in the Heaters tab, or both? I've already added the Solar Valve to the Valves tab, and the Pump speed for the Solar Circuit to the Pumps tab. The problem is when I tap the Solar feature on the dashPanel home page, all that happens is that the pump speed changes to the correct solar speed setting. But the valve does not rotate.

I have all delays unchecked just to make sure those are not in effect.

Am I thinking about this right? I'm sure its just another one of those obvious things that I'm missing. BTW, I did read all the posts and [wiki pages](https://github.com/tagyoureit/nodejs-poolController/wiki/Nixie-Controller-Setup#solar-heater) that mention solar, but I'm still stuck. 

Thank you kindly once again for your patience.

![solar0](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/c9fa01a4-f8b1-42d3-8175-79b589bcbe8a)

![solar1](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/7ae63c9a-62dd-4992-ad35-1b666170cf86)

![solar2](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/b4838b87-5a7a-46df-b5d1-998f824c4db2)

![solar3](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/f3bff0d8-951a-41cd-8517-bf4e04b5fd4a)

![solar4](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/7ab135b7-3d69-41b0-b1a4-3c3c6e492265)

**Comment by rstrouse** [comment]
To set up solar heating you need to add a valve and then add the heater.  You do not need or want a feature.  Once you add the heater it will add a virtual circuit into the configuration that will trigger whenever Nixie calls for solar heat.  This virtual circuit is what you assign to the pump and the valve.

To control the heater, click on the circled area and a popover will appear that allows to to set the target temperature.  Nixie will take care of calculating the deltas and ensure there is enough heat in the panels to heat the water.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/07f05319-6179-42f3-9733-3af8ec06eae0)

The solar temp probe is important and you will want one.  However, you can create another feed on your air temp sensor to feed it to the solar temp.  While this is not ideal it will at least not call for heat when there is none to be had.

**Reply by CodeSlinger69** [reply]
Thank you, that was the paradigm shift that I was missing. When njsPC/dashPanel is hooked up to a Pentair OCP, there is a Circuit/Feature button for "Solar" that automagically shows up. This button simply rotates the solar valve. In Nixie, you have more flexibility, rather than just on/off - turn on the solar, but only until the water temp hits a certain point (the "Set Point"). Nixie is better of course, as there was no obvious way with the OCP to have the solar automatically turn off when the target temperature was reached.

I was thrown off by the dashPanel interface a bit too, fwiw. The other interactable elements are mostly obvious on the dashPanel; i.e. they are clearly buttons. The red box you outlined above does not intuitively indicate that it will do something when tapped, so I hadn't even considered tapping on it.

Some of the other interface elements show up as buttons that are not tappable, like the virtual circuits, but at least those look like buttons so you can try tapping on them.

I have a solution for the solar temp sensor and I'll write it up in my other thread.

**Comment by CodeSlinger69** [comment]
@rstrouse, as I have been working through my understanding of this part of Nixie, I have a follow-up question on the logic in the solar heating algorithm. And maybe I've discovered a logic error, but please correct me if I am misunderstanding.

Let me lay out my thinking. Sorry this is kind of long.

I first read [this](https://github.com/tagyoureit/nodejs-poolController/wiki/Nixie-Controller-Setup#solar-heater) very carefully to understand the logic flow that Nixie uses to determine whether solar heating or cooling is needed. Key concept: "When the water temperature falls below the setpoint and the water temperature plus the Start Temp Delta is greater than the setpoint then Nixie will trip the solar valve relay to begin heating."

However, this doesn't quite match up to the algorithm in the [code](https://github.com/tagyoureit/nodejs-poolController/blob/e0c178df3a4cf25d09e70537bea7757beb6cbe0d/controller/boards/SystemBoard.ts#L4267). The important parts with English below:

`if (body.temp < cfgBody.heatSetpoint &&`
If the current water temperature is less than our desired temperature (a.k.a setpoint, heat point)
AND

`state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta))`
the current solar temperature is greater than the current water temperature plus either the "Stop Temp Delta" if the solar circuit is currently on or the "Start Temp Delta" if the solar circuit is currently off
THEN

start/continue using the solar heating circuit.

Example assuming the solar circuit is currently off: My current water temp is 80, the desired water temp is 85, the current solar temp is 110, and my Start Temp Delta is 6 (the default):
80 < 85 (YES!)
110 > 80 + 6 (YES!)

Once the solar circuit starts running the math is:
80 < 85 (YES!)
110 > 80 + 3 (YES!) (3 is the default Stop Temp Delta)

My first observation: Based on this analysis of the algorithm the way it is coded right now, I think the key concept in the Wiki page should instead state it like this: "When the water temperature falls below the setpoint and the solar temperature is greater than the setpoint plus the Start Temp Delta then Nixie will trip the solar valve relay to begin heating." 

My second observation: Consider the situation where the water temp is at our target of 85, but drops to 84. Nixie will start the circuit, assuming enough solar heat. Then it will shut if off when the water temp gets back to 85. This could cause relatively rapid cycling of the solar circuit on and off. The section of code right after this is for ultratemp heating logic, and the comments state that Pentair's OCPs consider a "temperature differential" that needs to be exceeded before the heater starts, just like home HVAC systems operate. Should this temp differential logic be added to the solar heating logic?

I plan on taking advantage of nocturnal cooling, due to where I live, so I looked at that logic too, to make sure I understand it. There may be a logic bug. Here's the [code](https://github.com/tagyoureit/nodejs-poolController/blob/e0c178df3a4cf25d09e70537bea7757beb6cbe0d/controller/boards/SystemBoard.ts#L4273):

`body.temp > cfgBody.coolSetpoint &&`
Current water temp is greater than our desired temperature (a.k.a setpoint, cool point)
AND

`state.heliotrope.isNight &&`
its night time
AND

`state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta)`
the current solar temperature is greater than the current water temperature plus either the "Stop Temp Delta" if the solar circuit is currently on or the "Start Temp Delta" if the solar circuit is currently off

My third observation: Shouldn't this be a less than operator? We don't want to cycle water through the solar circuit to cool the pool unless the solar temp is less than the water temp, right? 

My fourth observation: The Wiki page confirms that "less than" is the intent: "Nocturnal Cooling: When enabled Nixie will engage the solar valve if the Solar Temperature sensor reads less than the high limit setting for the body and the calculated sunset time has expired." However, if that Wiki sentence is correct, shouldn't the logic use cfgBody.coolSetpoint rather than body.temp?

Thanks again for listening!

**Comment by rstrouse** [comment]
WRT to the wiki you are right it probably needs some adjustment and I will go back through it to make sure it matches how it is coded.  However, when it comes to solar heaters the heuristics are quite a bit different from a temp differential that you get from a high BTU heater.  I will look at them again but they are probably correct to both ensure the heating seeks the setpoint and it doesn't short cycle.

On the cooling side at first glance I would think you are correct.  I need to go back and look at the nocturnal cooling parameters but in the end it should be looking at the coolSetpoint, the current body temperature, and the temperature at the roof.  You do not want the water coming off the roof until it has sufficiently cooled.

**Comment by CodeSlinger69** [comment]
I worked on this a bit more, and almost got my pool to automatically turn on nocturnal cooling last night. This seems to be better than the above:

```
else if (heater.coolingEnabled && body.temp > cfgBody.coolSetpoint && state.heliotrope.isNight &&
         state.temps.solar < body.temp - (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta)) {
```

Notice the subtraction operator rather than the addition operator. I'll see if I can update my pull request accordingly. 
Edit: I can't seem to figure out how to update the pull/branch, so I just pasted the complete conditional statement into the pull request discussion thread.

**Comment by CodeSlinger69** [comment]
I still aspire to automatic nocturnal cooling, so I'm working on this a bit more. A follow-up question on how to add a specific pump speed to support the solar function. As you can see below, I added a solar heater, but on the pumps page, there are many choices for the 2400 speed that I need. Through trial and error, selecting any of the solar circuits will not move the pump to 2400rpm when the solar heating/cooling circuit turns on. However, selecting "Pool Heater" for the solar pump speed WILL set the pump at 2400 when the solar circuit is turned on. Is this expected?

Secondly, the moment I added the solar heater on the Heaters tab, I got 5 new virtual circuits related to heating as shown below. Is this also expected?

![solar1](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/a76265d8-aace-4b54-b13f-fe8a80d5282e)

![solar2](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/08b92ea3-a99c-4d52-80a6-fb115eea21c2)

![solar3](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/2cb30bc7-8c04-42d2-9361-c37cf74886cc)

**Reply by rstrouse** [reply]
No remove that relay from the definition of the heater.  In your case it will be blank. 

You want to create a valve and assign it there.  Assign the circuit for the valve to `Solar`.  You will use this virtual circuit for the pump as well.  In your instance you only have one body so `Solar` and `Pool Solar` are identical but use solar because it has additional logic form managing delays.  

`Heater` and `Pool Heater` is for gas, electric, heatpumps, or hybrid heaters that may be engaged.  If you do not have these then they will never engage but they are there to handle the underlying logic for economy heating.  It's cheaper to heat with the sun than firing up gas or electric.

**Reply by CodeSlinger69** [reply]
Great, thanks. Heater, valve, and pump set up as you described. Let’s see if it turns on now for cooling when the weather allows.

**Comment by CodeSlinger69** [comment]
Well, I'm trying to work through another setup/interface question. 

What is the proper Nixie way of setting solar heat/cool setpoints? In the schedule, or in the body panel on the main window? I poked around with all the various combinations of doing so, but I'm just not sure why there are two places to define these setpoints, and what the different functionality might be. So I'm not sure what the "right" way is to do this.

I have notices that if I use the schedule, then it overrides and replaces the heat setpoint in the body panel, but on the other hand the cool setpoint is not exposed in the schedule so it must be set in the body panel. Additionally, I'm seeing some strangeness in when these values get saved/overwritten, but let me first figure out the proper way to do it, as I may be just not using the tool correctly.

thx.

![heat1](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/8aed9a23-3ce9-4987-9f76-1d26dc652887)
![heat2](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/58a0f0f6-eda0-405c-8ddc-07f767793bf5)

**Reply by rstrouse** [reply]
With the updates that I am working on, Nixie will now also support the cool setpoint on the schedule.

**Reply by CodeSlinger69** [reply]
I'm testing these changes now, and I have to say I like it a lot. This makes more logical sense to me. You set the heat/cool setpoints in the schedule for automatic operation. But if you want to override the current schedule, you can tap on the body panel on the main screen and do a quick temporary adjustment (at least until the next schedule comes around).

Only thing I might suggest is setting the default cool setpoint to something else. It currently is set at 100F when you create a new schedule item. If my pool is warmer than 100F, and I want to cool it down to 100F, I have some serious problems. I might suggest 75? Right now in PHX, my pool is at 89F, and I'd like to cool it to less than 80F using nocturnal solar cooling, so I set it to 80 for now.

I've fed some temps through REM and njSPC using Postman, and nocturnal cooling does seem to engage as expected. 

Now to pray for some cooler nights. The temperature is not dropping below 90 at night right now, so I can't cool with solar anyway. But hopefully we'll get some cooler nights here in a month or so.

**Comment by CodeSlinger69** [comment]
Last night cooled off, finally. But I'm still trying to get this to work correctly. Would you mind taking a peek at the below and letting me know what I have misconfigured? In the first image, you can see that the nocturnal cooling is engaged (the little snowflake icon), but the pump was not turned up to the solar RPM. In the second image, you can see that the valve for solar is not engaged. The last three images show the config.

Thank you. Since I know how to use Postman now to simulate temps through REM, I can run some tests and capture logs if it would be helpful.

![solar1](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/6a6d0416-eaf5-491b-9d19-079d5e0b541e)

![solar2](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/6ff18a36-e1bb-4b17-95cb-dbd496798975)

![solar3](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/9dc40f50-523a-4b21-8dec-8745159310e6)

![solar4](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/b79b990c-a733-4898-9ca0-d49c51333011)

![solar5](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/57f4eed1-b8f0-4d26-a586-218df093340f)

**Reply by rstrouse** [reply]
Well it looks like it is not engaging the virtual circuit for the heater status when it is cooling.  I will have a look.

**Reply by rstrouse** [reply]
Pull njsPC.  I believe I got all the scenarios where the solar virtual circuit should be engaged.

**Reply by CodeSlinger69** [reply]
Thank you @rstrouse; pulled. I'll do some testing and come back here with my results.

**Reply by CodeSlinger69** [reply]
Found an issue, I think, not with this commit, but with one commit back. I hadn't pulled in a week, so hadn't tested your recent changes. In [this commit](https://github.com/tagyoureit/nodejs-poolController/compare/cac6d5229e91a957d64902c24b3d4ad1214bbacc...a8acb2ad70029d96559699e7b23e2d8e9968c0f9), there is this change in Schedule.ts, line 54:

`+                 if (!sscheds[i].isOn || sscheds[i].scheduleTime.shouldBeOn) continue;`

This caused none of my schedules to run, as seen below. I think the fix is this, but please check my work:

`+                 if (sscheds[i].isOn || !sscheds[i].scheduleTime.shouldBeOn) continue;`



![sched1](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/543c8f53-f4c9-4511-a7a4-05735197a45d)

**Reply by rstrouse** [reply]
Not quite right but you are in the correct place.  The code should read.

```javascript
                // We only care about schedules that are currently running or should be running.
                if (!sscheds[i].isOn && !sscheds[i].scheduleTime.shouldBeOn) continue;

```

**Reply by CodeSlinger69** [reply]
Thanks. Pulled, and the fix seems to be working. I'll get to nocturnal cooling testing later today, I think.

**Reply by CodeSlinger69** [reply]
I got around to testing nocturnal cooling, but still not engaging the right pump speed, or the solar valve, as seen below. I did git pull and compile prior to testing, just to be sure. Is there anything I can provide to help with this?

![noc1](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/44c8bca9-6cb5-4016-be09-e03e6f702e55)

![noc2](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/20e6515a-a682-4cd4-816a-d88098297f00)

**Reply by rstrouse** [reply]
When it is in this state I need a replay of the data.  The solar engagement for cooling should be tagged to the isCooling attribute on the solar heater.  There is a time delay between on and the cooling being engaged but it is not really all that long.  Max 30 seconds as it allows all the other elements to stabilize.

EDIT:  I'll change one of my test systems around so I can adjust the temp downward to see if the virtual circuit engagement is goofy.

**Reply by CodeSlinger69** [reply]
[removed]

**Comment by rstrouse** [comment, marked as answer]
Ok so I built a more complete test rig and found the condition that was failing.  Pull njsPC as it should now engage the Solar virtual circuit on a single body solar array with nocturnal cooling when the temperature rises above the setpoint.

**Reply by CodeSlinger69** [reply]
Yes, that did it, thank you so much. I realize that I am the only one using this feature, so I appreciate your patience as we worked through it together. It should be cool enough tonight that this will kick in and help with the pool temps. We'll see.

**Reply by rstrouse** [reply]
No worries.  I missed a couple of spots where the virtual circuits were being triggered.  It just took setting up your configuration to get it to reveal them.
